pkg/index/index.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package index
+
+import (
+	"container/list"
+	"errors"
+	"fmt"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+var (
+	ErrNodeNotFound = errors.New("Node not found")
+)
+
+type Node interface {
+	// Returns the string representation (usually the ID) of the node
+	String() string
+
+	// Returns the string respresentation of the nodes parent (usually it's ID)
+	Parent() string
+
+	// Deep copy of the node
+	Copy() Node
+}
+
+type node struct {
+	Node
+	parent   *node
+	children []*node
+	mask     uint32
+}
+
+func (n *node) addChild(child *node) {
+	n.children = append(n.children, child)
+}
+
+type Index struct {
+	root        *node
+	lookupTable map[string]*node
+	l           sync.Mutex
+}
+
+func NewIndex() *Index {
+	i := &Index{
+		lookupTable: make(map[string]*node),
+	}
+
+	return i
+}
+
+// Insert inserts a copy of the given node to the tree under the given parent.
+func (i *Index) Insert(n Node) error {
+	i.l.Lock()
+	defer i.l.Unlock()
+
+	_, ok := i.lookupTable[n.String()]
+	if ok {
+		return fmt.Errorf("node %s already exists in index", n.String())
+	}
+
+	log.Debugf("Inserting %s (parent: %s) in index", n.String(), n.Parent())
+
+	newNode := &node{
+		Node: n.Copy(),
+	}
+
+	if n.Parent() == n.String() {
```

Comment:
Parent can't return the `Node` because our `Node` interface is used to keep string references.  The `Index` uses these to build the tree using actual pointers ( to`node`s).